runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/MemoryPoolAssigner.java

```diff
@@ -0,0 +1,170 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.nemo.conf.JobConf;
+
+import javax.inject.Inject;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * The MemoryPoolAssigner assigns the memory that Nemo uses for writing data blocks from the {@link MemoryPool}.
+ * Memory is represented in chunks of equal size. Consumers of off-heap memory acquire the memory by requesting
+ * a number of {@link MemoryChunk} they need.
+ *
+ * MemoryPoolAssigner currently supports allocation of off-heap memory only.
+ *
+ * The MemoryPoolAssigner pre-allocates user-defined amount of memory at the start.
+ * More memory can be allocated on-demand, but if there is no more memory to allocate, MemoryAllocationException
+ * is thrown and the job fails. // TODO #397: Separation of JVM heap region and off-heap memory region
+ */
+@ThreadSafe
+public class MemoryPoolAssigner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MemoryPoolAssigner.class.getName());
+
+  private final int chunkSize;
+
+  private static final int MIN_CHUNK_SIZE_KB = 4;
+
+  private final MemoryPool memoryPool;
+
+  @Inject
+  public MemoryPoolAssigner(@Parameter(JobConf.MaxOffheapMb.class) final int maxOffheapMb,
+                            @Parameter(JobConf.ChunkSizeKb.class) final int chunkSizeKb) {
+    if (chunkSizeKb < MIN_CHUNK_SIZE_KB) {
+      throw new IllegalArgumentException("Chunk size too small. Minimum chunk size is 4KB");
+    }
+    final long maxNumChunks = (long) maxOffheapMb * 1024 / chunkSizeKb;
+    if (maxNumChunks > Integer.MAX_VALUE) {
+      throw new IllegalArgumentException("Too many pages to allocate (exceeds MAX_INT)");
+    }
+    if (maxNumChunks < 1) {
+      throw new IllegalArgumentException("The given amount of memory amounted to less than one chunk.");
+    }
+    this.chunkSize = chunkSizeKb * 1024;
+    this.memoryPool = new MemoryPool((int) maxNumChunks, this.chunkSize);
+  }
+
+  /**
+   * Returns a single {@link MemoryChunk} from {@link MemoryPool}.
+   *
+   * @return a MemoryChunk
+   * @throws MemoryAllocationException if fails to allocate MemoryChunk.
+   */
+  public MemoryChunk allocateChunk() throws MemoryAllocationException {
+    return memoryPool.requestChunkFromPool();
+  }
+
+  /**
+   * Returns all the MemoryChunks in the given List of MemoryChunks.
+   *
+   * @param target  the list of MemoryChunks to be returned to the memory pool.
+   */
+  public void returnChunksToPool(final List<MemoryChunk> target) {
+    for (final MemoryChunk chunk: target) {
+      memoryPool.returnChunkToPool(chunk);
+    }
+  }
+
+  /**
+   * Returns the chunk size of the memory pool.
+   *
+   * @return the chunk size in bytes.
+   */
+  public int getChunkSize() {
+    return chunkSize;
+  }
+
+  @VisibleForTesting
+  /**
+   * Returns the number of chunks in the pool.
+   */
+  public int returnPoolSize() {
```

Comment:
Remove `public`, to make it package-private.
The tests are under the same package.

poolSize()?